File: problem_021.lua
#!/bin/env lua

print("Project Euler - Problem 21:")
print("Evaluate the sum of all the amicable numbers under 10000.\n")

limit = 10000

-- Initialize array
sums = {0}
for i=1, limit do
   sums[i] = 0
end

-- Calculate divisor sums
for i = 1, limit do
   n = i + i
   while n < limit do
      sums[n] = sums[n] + i
      n = n + i
   end
end

-- Check if amicable
sum = 0
for i = 2, limit do
   if sums[i] < i then
      if sums[sums[i]] == i then
         sum = sum + i + sums[i]
      end
   end
end

print(sum)

File: problem_021.js
#!/usr/bin/node

console.log("Project Euler - Problem 20:");
console.log("Evaluate the sum of all the amicable numbers under 10000.\n");

const limit = 10000;
var numbers = [];
var sum = 0;

// Initialize array
for (var n=0; n<limit; n++)
    numbers.push(1);

// Find all divisor sums
for (var n=2; n<limit; n++){
    let i = n + n;
    while (i < limit) {
	numbers[i] += n;
	i += n;
    }

    // Record amicable numbers
    if (numbers[n] < n && numbers[numbers[n]] == n)
	sum += n + numbers[n]
}

console.log(sum);

